services/src/main/java/org/keycloak/protocol/oidc/mappers/AbstractUserRoleMappingMapper.java

```diff
@@ -90,11 +87,11 @@ private static void mapClaim(IDToken token, ProtocolMapperModel mappingModel, Ob
         }
 
         if (clientId != null) {
-            // case when clientId contains dots
-            clientId = DOT_PATTERN.matcher(clientId).replaceAll(DOT_REPLACEMENT);
             Matcher matcher = CLIENT_ID_PATTERN.matcher(protocolClaim);
             if (matcher.find()) {
-                protocolClaim = matcher.replaceAll(clientId);
+                // case when clientId contains backslash, dot or dollar signs
+                String replacement = clientId.replace("\\", "\\\\\\\\").replace(".", "\\\\.").replace("$", "\\$");
+                protocolClaim = matcher.replaceAll(replacement);
```

Comment:
```suggestion
                protocolClaim = matcher.replaceAll(Matcher.quoteReplacement(clientId));
```

Reply:
This one is not so easy, because the `clientId` should escape first dots and backslashes for the claim itself. I added the `quoteReplacement` but it's:

```
protocolClaim = matcher.replaceAll(Matcher.quoteReplacement(clientId.replace("\\", "\\\\").replace(".", "\\.")));
```

Reply:
Thank you for the update. I now know that I still have to learn more about claims. As the test parses it in the response, I approve this. It would still be good if someone else maybe from @keycloak/core would have a look.